# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Создадим функцию для вывода сводных таблиц:

In [22]:
def dependence(row):
    result = data.groupby(row)['debt'].agg(['count', 'sum', 'mean']).sort_values(by='count', ascending=False)
    result['mean'] = result ['mean'].apply(lambda x: "%.2f%%" % (x * 100))
    return result

In [23]:
dependence('children')

count   sum   mean
children                    
0         14091  1063  7.54%
1          4808   444  9.23%
2          2052   194  9.45%
3           330    27  8.18%
4            41     4  9.76%
5             9     0  0.00%

**Вывод:** Исходя из вышеизложенного можно сделать вывод, что заемщики без детей выплачивают кредит исправнее. У семей с 1-м и 2-мя детьми процент просрочек выше, вероятно это связано с дополнительной финансовой нагрузкой. А вот данных о многодетных заемщиках недостаточно, их учитывать нецелесообразно.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
dependence('family_status')

count  sum   mean
family_status                           
женат / замужем        12261  927  7.56%
гражданский брак        4134  385  9.31%
Не женат / не замужем   2796  273  9.76%
в разводе               1189   84  7.06%
вдовец / вдова           951   63  6.62%

**Вывод:** Заемщики, у которых есть или был официальный брак, более ответственны к выплатам по кредиту.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [25]:
data['total_income_category'] = data['total_income_category'].map({'E':'E_0_30тыс','D':"D_30_50тыс",'C':'C_50_200тыс', 'B':'B_200_1млн', 'A':'A_1млн' })

In [26]:
dependence('total_income_category')

count   sum   mean
total_income_category                    
C_50_200тыс            15921  1353  8.50%
B_200_1млн              5014   354  7.06%
D_30_50тыс               349    21  6.02%
A_1млн                    25     2  8.00%
E_0_30тыс                 22     2  9.09%

**Вывод:** Данных для анализа достаточно у категорий В и С, сравним эти 2 категории. Заемщики с доходом от 200 тысяч показали хороший результат к выплатам в срок. На 2-м месте заемщики с доходом 50-200 тысяч, эта категория клиентов чаще остальных берет кредиты, вероятно, в виду отсутствия финансового резерва.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [27]:
dependence('purpose_category')

count  sum   mean
purpose_category                           
операции с недвижимостью  10751  780  7.26%
операции с автомобилем     4279  400  9.35%
получение образования      3988  369  9.25%
проведение свадьбы         2313  183  7.91%

**Вывод:** Заемщики категории "операции с недвижимостью" занимают лидирующую позицию по количеству и по выплатам в срок.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски были обнаружены в данных о стаже и доходах. Причинами появления пропусков может быть ошибка при вводе данных, сбой в работе БД.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медианное значение более устойчиво к выбросам и аномальным значениям. 
К примеру, сравним среднее значение и медианное столбца 'days_employed', показатели сильно отличаются, в столбце присутствуют аномальные значения. Таким образом, с помощью медианы можно получить более точные данные и правильнее интерпретировать статистику.

In [28]:
print ('Среднее значение общего трудового стажа в днях =', data['days_employed'].mean())
print ('Медианное значение общего трудового стажа в днях =', data['days_employed'].median())

Среднее значение общего трудового стажа в днях = 67148.97213258488
Медианное значение общего трудового стажа в днях = 1996.732855437927


#### 3.7 Дополнительно проверим как влияет наличие высшего образования на возврат кредита в срок.

Создадим сводную таблицу с помощью метода pivot_table(): 

In [29]:
data_pivot = data.pivot_table(index='education', values='debt', aggfunc={'count','sum','mean'})
data_pivot['mean'] = data_pivot['mean'].apply(lambda x: "%.2f%%" % (x * 100))
data_pivot.sort_values(by='count',ascending=False)

count    mean     sum
education                                   
среднее              15075.0   8.99%  1355.0
высшее                5227.0   5.32%   278.0
неоконченное высшее    741.0   9.18%    68.0
начальное              282.0  10.99%    31.0
ученая степень           6.0   0.00%     0.0

**Вывод:** Однозначно можно утверждать, что высшее образование положительно влияет на выплаты кредита в срок. Большинство заемщиков имеют среднее образование, поэтому наличие диплома не является определяющим фактором при выдаче кредита, но процент задержки у этой категории высокий.

### Шаг 4: общий вывод.

Для исследования надежности клиентов банка проведен анализ данных о их платежеспособности. 

На этапе предобработки данных были выявлены и заполнены пропуски в столбцах с информацией о доходах и трудовом стаже, обработаны аномальные значения в столбце с информацией о количестве детей и удалены дубликаты в столбце, содержащий информацию об уровне образования.

Проведена категоризация клиентов по уровню доходов и цели получения кредита.

В ходе анализа рассмотрели, как влияют на возврат кредита в срок:
- **количество детей в семье:**

*Категория клиентов без детей занимает 1 место по количеству и имеет самый низкий процент задержек - 7.54%, клиенты с высоким процентом задержек 9.45% имеют 2-х детей. В выборку не попали многодетные в виду их малого количества относительно других категорий.*
    
- **семейное положение:**

*Наименьший процент задержек по кредиту у категории вдов/вдовцов - 6.62%, далее по рейтингу категория разведенных - 7.06%. Чаще всего кредит берут клиенты в браке и занимают 3 место по выплатам в срок, процент задержек - 7.56%.*

*Самый высокий процент задержек 9.76% у клиентов в неофициальных отношениях.*

- **уровень дохода:**

*У клиентов с доходом от 200 тысяч до 1 миллиона наименьший процент задержек по кредиту - 7.06%, c доходом от 50 до 200 тысяч чаще обращаются за кредитами, процент задержек равен 8.50%.*

*Категория клиентов с низким и высоким значениями доходов не вошли в исследование, данных недостаточно.*

- **цель получения кредита:**

*Большинство клиентов берут кредиты на жилье, эта категория имеет самый низкий процент задержек - 7.26%. Не торопятся выплачивать категория клиентов с кредитами на образование (9.25%) и автокредитами (9.35%).*


**Исходя из вышеизложенного можем сделать вывод:**

**Надежный клиент** - клиент без детей в официально зарегистрированном браке, либо в разводе с доходом от 200 тысяч. 

**Менее надежный клиент** - клиент в неофициальных отношениях, либо свободный с 1-м или 2-мя детьми и доходом от 50 до 200 тысяч. 

Клиенты ответственее подходят к выплатам при оформлении кредитов на недвижимость. Также благоприятно сказывается на выплаты наличие высшего образования, процент задержек сравнительно низкий - 5.32%. 

В данном анализе по некоторым категориям были малочисленные выборки, которые нецелесообразно учитывать. 
